### PR TITLE
Backporting (s3-output PR #249) the fix of uploading corrupted GZIP to AWS S3 object.

### LINKED ISSUE #16: GZIP files get corrupted and uploaded to S3 when Logstash abnormally stopped.
### Issue description
When logstash is shut down abnormally (eg.: receives a kill signal), the s3 files created in the temporary directory are saved, and automatically uploaded on next startup (when the restore option is set to true). However, when we download and check the uploaded S3 file, it is noticed that files are corrupted.

See original issue: https://github.com/logstash-plugins/logstash-output-s3/issues/189

We implemented and release this feature with `logstash-output-s3` plugin but didn't backport yet since they manage the source separately. This issue also covers the this https://github.com/logstash-plugins/logstash-integration-aws/issues/15

### Tasks needs to be done
1. Integration AWS plugin is based on pure Ruby but our logic introduced in Java. (We are moving to JVM world for the long term). We need to setup the JVM environment with the plugin.
2. Follow up the logics introduced in output S3 plugin and test carefully all the cases.

Note that, we already considered several possible solutions and selected the best one by the discussion in a team. No need to further research in case of solution. Just need to copy what approach/code S3 output is using.

### PR DESCRIPTION
**Pull request by mashhurs**
## Release notes

Logstash now restores and uploads the corrupted files to S3.

## What does this PR do?
Backports `logstash-output-s3` corrupted S3 file issue case ([PR #249](https://github.com/logstash-plugins/logstash-output-s3/pull/249)).

## Why is it important/What is the impact to the user?
When using GZIP encoding option with output to AWS S3 plugin, there are cases where Logstash may be crashed. When Logstash crashed GZIP stream is left opened and no tail in the file exist. Logstash uploads corrupted file to S3 at restart but customers who download S3 file and use, they figured out the file is corrupted.
This PR aims to recover the corrupted file at restart time and upload healthy GZIP file to S3.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files (and/or docker env variables)
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ]

## How to test this PR locally
1. Use following config
```
input {
  stdin {}
}
output {
    s3 {
        region => "ca-central-1"
        bucket => "logstash-mashhur-test-1"
        codec => "json_lines"
        canned_acl => "private"
        prefix => "test-%{+YYYY.MM.dd}"
        additional_settings => {
            "force_path_style" => true
        }
        encoding => "gzip"
        upload_queue_size => 10
        upload_workers_count => 2
        time_file => 1
        rotation_strategy => "time"
        temporary_directory => "/Users/mashhur/Dev/elastic/temp/s3-temp"
        validate_credentials_on_root_bucket => false
    }
}
```
2. Send some input data and kill the Logstash process 
3. Re-run the Logstash
4. Download S3 object uploaded with AWS-CLI. Note that downloading on browser automatically unzips the file. Better use plain AWS-CLI.
```
aws s3api get-object --bucket logstash-mashhur-test-1 --key test-2023.01.05/ls.s3.b2854743-82b4-4a73-9350-36746e6ff3aa.2023-01-05T13.21.part2.txt.gz my_downloaded_s3_object.txt.gz
```
5. Unzip the file and check the content
```
gunsip my_downloaded_s3_object.txt.gz
cat my_downloaded_s3_object.txt
```

## Related issues
- Closes #16

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,6 @@
+## 7.1.0
+  - Plugin restores and uploads corrupted GZIP files (caused by abnormal termination) to AWS S3 [#N](https://github.com/logstash-plugins/logstash-integration-aws/pull/N)
```
Comment (mashhurs) [inline CHANGELOG.md:2]:
```suggestion
  - Plugin restores and uploads corrupted GZIP files (caused by abnormal termination) to AWS S3 [#20](https://github.com/logstash-plugins/logstash-integration-aws/pull/20)
```

File: logstash-integration-aws.gemspec [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Gem::Specification.new do |s|
   s.name            = "logstash-integration-aws"
-  s.version         = "7.0.1"
+  s.version         = "7.1.0"
```
Comment (robbavey) [inline logstash-integration-aws.gemspec:3]: Nit: 

As we have a `VERSION` file, we should refer to it here too, to avoid a mismatch

**Review by robbavey — APPROVED** [review]
LGTM - one request to use the `VERSION` file in gemspec as well as in gradle.

File: logstash-integration-aws.gemspec (hunk 2 above)
Comment (mashhurs) [reply logstash-integration-aws.gemspec:3]: Good catch, thanks.
I think it is better to load the version from the `VERSION` file.

File: logstash-integration-aws.gemspec [hunk 3]
```diff
@@ -1,6 +1,8 @@
+INTEGRATION_AWS_VERSION = File.read(File.expand_path(File.join(File.dirname(__FILE__), "VERSION"))).strip unless defined?(INTEGRATION_AWS_VERSION)
+
 Gem::Specification.new do |s|
   s.name            = "logstash-integration-aws"
-  s.version         = "7.0.1"
+  s.version         = INTEGRATION_AWS_VERSION
```
Comment (mashhurs) [inline logstash-integration-aws.gemspec:5]: @robbavey how do you think for this change?

**Review by robbavey — APPROVED** [review]
LGTM

**Pull request merged**